common/src/patch.rs

```diff
@@ -0,0 +1,165 @@
+    /// branch.
+    pub merge_base: Option<Oid>,
+}
+
+impl Metadata {
+    pub fn state(&self) -> State {
+        if self.merge_base.unwrap() == self.commit {
+            return State::Merged;
+        }
+        State::Open
+    }
+
+    pub fn name(&self) -> String {
+        str::replace(&self.id, "radicle-patch/", "")
+    }
+}
+
+/// List patches on the local device. Returns a given peer's patches or this peer's
+/// patches if `peer` is `None`.
+pub fn list<S>(
+    storage: &S,
+    peer: Option<PeerId>,
+    project: &project::Metadata,
+    repo: &git2::Repository,
+) -> Result<Vec<Metadata>, Error>
+where
+    S: AsRef<ReadOnly>,
+{
+    let storage = storage.as_ref();
+    let mut patches: Vec<Metadata> = vec![];
+    let master = repo
+        .resolve_reference_from_short_name(&format!("rad/{}", project.default_branch))?
+        .target()
+        .unwrap();
+
+    if let Ok(refs) = Refs::load(&storage, &project.urn, peer) {
+        let blobs = match refs {
+            Some(refs) => refs.tags().collect(),
+            None => vec![],
+        };
+        for blob in blobs {
+            let object = storage.find_object(blob.1)?.unwrap();
+            let tag = object.peel_to_tag()?;
+            let merge_base = repo.merge_base(master, tag.target_id())?;
+
+            patches.push(Metadata {
+                id: tag.name().unwrap().to_string(),
+                peer: peer.unwrap_or(*storage.peer_id()),
+                message: Some(tag.message().unwrap().to_string()),
+                commit: tag.target_id(),
+                merge_base: Some(merge_base),
+            });
+        }
+    }
+
+    Ok(patches)
+}
+
+/// Create and push tag to monorepo.
+pub fn create(repo: &git2::Repository, message: &str, verbose: bool) -> anyhow::Result<()> {
+    let head = repo.head()?;
+    let current_branch = head.shorthand().unwrap_or("HEAD (no branch)");
+    let patch_tag_name = format!("radicle-patch/{}", current_branch);
+    let spinner = term::spinner("Creating patch.");
+
+    match git::add_tag(repo, message, &patch_tag_name) {
+        Ok(patch_tag_name) => patch_tag_name,
+        Err(err) => {
+            spinner.failed();
+            return Err(err);
+        }
+    };
+    if verbose {
+        term::info!("Created tag {}", &patch_tag_name);
+    }
+
+    let output = match git::push_tag(&patch_tag_name) {
+        Ok(output) => output,
+        Err(err) => {
+            spinner.failed();
+            return Err(err);
+        }
+    };
+    if verbose {
+        term::blob(output);
+    }
+    spinner.finish();
+
+    Ok(())
+}
+
+/// Adds patch details as a new row to `table` and render later.
+pub fn print<S>(
+    storage: &S,
+    peer: Option<PeerId>,
+    project: &project::Metadata,
+    patch: &Metadata,
+    table: &mut term::Table<2>,
+) -> anyhow::Result<()>
+where
+    S: AsRef<ReadOnly>,
+{
+    let storage = storage.as_ref();
+    let peer_id = peer.unwrap_or(*storage.peer_id());
```

Comment:
Could we instead use the `peer_id` in `Metadata`? Kind of strange to pass in a peer into this function, unless I'm missing something.

Reply:
Changed the function and it's signature quite a bit.